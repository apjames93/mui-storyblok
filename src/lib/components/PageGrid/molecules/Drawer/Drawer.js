import React, { Suspense, lazy } from 'react';
import PropTypes from 'prop-types';
import { Drawer as MuiDrawer } from '@material-ui/core';
import {
  validComponents,
  muiStringProp,
} from 'lib/utils/customProps';
import { renderComponentsWithBridge } from 'lib/utils/customComponents';
import StoryBlok from 'lib/utils/Storyblok';

const List = lazy(() => import('lib/components/PageGrid/organisms/List/List'));

const components = {
  List,
};

const Drawer = ({
  anchor,
  rootClass,
  elevation,
  variant,
  toggleDrawer,
  content,
  open,
  dataBlokC,
  dataBlokUid,
  storyblokClass,
}) => {
  let width = '100%';
  if (anchor === 'left' || anchor === 'right') width = '32vw';
  const styles = StoryBlok.arrayToMuiStyles(rootClass, { minWidth: width });

  // eslint-disable-next-line no-multi-assign
  const handleToggleDrawer = (event) => {
    if (event.type === 'keydown' && (event.key === 'Tab' || event.key === 'Shift')) {
      return;
    }
    toggleDrawer(!open);
  };

  return (
    <MuiDrawer
      anchor={anchor}
      classes={{ paper: styles.root }}
      open={open}
      elevation={+elevation}
      variant={variant}
      onClose={e => handleToggleDrawer(e)}
      className={`${styles.root} ${storyblokClass}`}
      data-blok-c={dataBlokC}
      data-blok-uid={dataBlokUid}
    >
      {content.map((component, key) => (
        <Suspense fallback={<></>}>
          {renderComponentsWithBridge(components, { ...component, toggleDrawer }, key)}
        </Suspense>
      ))}
    </MuiDrawer>
  );
};

export default Drawer;
 Drawer.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui props: 'bottom', 'left', 'right', 'top'
   * Side from which the drawer will appear.
   */
  anchor(props, propName, componentName) {
    const validProps = ['bottom', 'left', 'right', 'top'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * number
   * The elevation of the drawer.
   * storyblok type is a number
   */
  elevation: PropTypes.string,
  /**
   * mui props: 'permanent', 'persistent', 'temporary'
   * The variant to use.
   */
  variant(props, propName, componentName) {
    const validProps = ['permanent', 'persistent', 'temporary'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * components:
   * MuiList
   */
  content(props, propName, componentName) {
    const components = ['MuiList', 'MuiListDropdown'];
    return validComponents(props, propName, componentName, components);
  },
};
 Drawer.defaultProps = {
  anchor: 'left',
  elevation: 16,
  variant: 'temporary',
  rootClass: [],
  content: [],
};
