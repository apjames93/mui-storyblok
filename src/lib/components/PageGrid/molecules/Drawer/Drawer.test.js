import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import Drawer from './Drawer';

function setup(open = false) {
  const props = {
    open,
    toggleDrawer: jest.fn(),
    _editable: '<!--#storyblok#{"name": "Drawer", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    elevation: 16,
    variant: 'temporary',
    content: [
      {
        _editable: '<!--#storyblok#{"name": "List", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'List',
        content: [
          {
            _editable: '<!--#storyblok#{"name": "ListItem", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
            component: 'ListItem',
            listItemText: [
              {
                _editable: '<!--#storyblok#{"name": "ListItemText", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
                primary: 'apples',
                component: 'ListItemText',
              },
            ],
          },
        ],
      },
    ],
  };
  const comp = mount(<Drawer {...props} />);
  return { comp, props };
}

describe('<Drawer />', () => {
  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<Drawer {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render Drawer', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });


  it('should handleClose and toggle drawer', () => {
    const { comp } = setup(true);

    expect(comp.find('WithStyles(ForwardRef(Drawer))').first().prop('open')).toEqual(true);
    const closeModal = comp.find('ForwardRef(Modal)').first().prop('onClose');
    act(() => {
      closeModal({ type: 'keydown', key: 'Tab' });
    });
    expect(comp.find('WithStyles(ForwardRef(Drawer))').first().props().open).toEqual(true);
    act(() => {
      closeModal({ type: 'keydown', key: 'Shift' });
    });
    expect(comp.find('WithStyles(ForwardRef(Drawer))').first().props().open).toEqual(true);
    act(() => {
      closeModal({ target: {} });
    });
    setTimeout(() => {
      expect(comp.find('WithStyles(ForwardRef(Drawer))').first().props().open).toEqual(false);
    }, 800);
  });
});
