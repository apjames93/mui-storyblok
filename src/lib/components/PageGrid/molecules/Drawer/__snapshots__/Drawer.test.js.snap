// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> snapshot 1`] = `null`;
